internal/converter/internal/otelcolconvert/envprovider/env_provider.go

```diff
@@ -0,0 +1,39 @@
+package envprovider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+)
+
+const SchemeName = "env"
+
+// provider is a custom environment variable provider
+type provider struct{}
+
+// NewFactory returns a custom environment provider factory
+func NewFactory() confmap.ProviderFactory {
+	return confmap.NewProviderFactory(newProvider)
+}
+
+func newProvider(confmap.ProviderSettings) confmap.Provider {
+	return &provider{}
+}
+
+func (s *provider) Retrieve(_ context.Context, val string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	if !strings.HasPrefix(val, s.Scheme()+":") {
+		return nil, fmt.Errorf("%q environment variable scheme is not supported by %q provider", val, s.Scheme())
+	}
+
+	return confmap.NewRetrieved(fmt.Sprintf("$${%s}", val))
```

Comment:
What is $$ mean in this context?

Reply:
I'll add a comment to the code there. That's the OTel dollar sign escape syntax. If I leave the incoming `$` untouched, the OTel parser will infinitely recurse trying to expand it out, so I escape it, which it converts downstream to `$` instead of re-taking the path of my new provider. Basically making OTel think it's just text instead of an envvar reference.